packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -284,6 +287,61 @@ export function buildImportAnalysisPlugin(config: ResolvedConfig): Plugin {
         return [url, resolved.id]
       }
 
+      // when wrapping dynamic imports with a preload helper, Rollup is unable to analyze the
+      // accessed variables for treeshaking. This below tries to match common accessed syntax
+      // to "copy" it over to the dynamic import wrapped by the preload helper.
+      const dynamicImports: Record<
+        number,
+        { declaration?: string; names?: string; chains?: string }
+      > = {}
+
+      if (insertPreload) {
+        let match
+        while ((match = dynamicImportTreeshakenRE.exec(source))) {
+          /* handle `const {foo} = await import('foo')`
+           *
+           * match[1]: `const {foo} = await import('foo')`
+           * match[2]: `const`
+           * match[3]: `{foo}`
+           * import end: `const {foo} = await import('foo')_`
+           *                                               ^
+           */
+          if (match[1]) {
+            dynamicImports[dynamicImportTreeshakenRE.lastIndex] = {
+              declaration: `${match[2]} ${match[3]}`,
+              names: match[3]?.trim(),
+            }
+            continue
+          }
+
+          /* handle `(await import('foo')).foo`
+           *
+           * match[4]: `(await import('foo')).foo`
+           * match[5]: `.foo`
+           * import end: `(await import('foo'))`
+           *                                  ^
+           */
+          if (match[4]) {
+            dynamicImports[
+              dynamicImportTreeshakenRE.lastIndex - match[5]?.length - 1
+            ] = { chains: match[5] }
```

Comment:
Does it make sense to make this:

```js
const names = match[5].match(/\.([^.?]+)/)?.[1] || ''
... = { declarations: `const {${names}}`, names: : `{ ${names} }`  }
```

The regex is borrowed from line 386, which maybe could be simplified and extracted.

This way all cases are consistently typed?

Reply:
I am afraid that we have to preserve the whole chaining `.foo` so that we can copy it to the transformed code. And it's a bit different between transforming `(await import('foo')).foo` and the other two statements.

Reply:
Yeah, that would be better 👍